Specify Account Management API

louischan-oursky opened the pull request:
ref DEV-1267

The motivation of Account Management API is to support adding a OAuth provider account WITHOUT using our existing settings page.

chpapa commented on docs/specs/account-management-api.md:
````diff
@@ -0,0 +1,861 @@
+
+```json
+{
+  "result": {
````
Why need this result field? Does this API prepare to support batch requests? If not this is redundant.

louischan-oursky replied:
The developer can first look at `error` or `result` in the API response. If `error` exists, then they can parse the error inside `error`, otherwise, they can parse the result inside `result`.

chpapa commented on docs/specs/account-management-api.md:
```diff
@@ -0,0 +1,861 @@
+      "identification": "username",
+      "login_id": "johndoe",
+      "claims": {
+        "preferred_username": "johndoe"
```
actually in what cases would the `claims` be different from the `login_id`?

louischan-oursky replied:
No case. Just add `claims` so that the developer can just read from the `claims` to get the `email`, `phone_number`, etc. Otherwise, they have to switch on `identification` and determine the login ID is actually a email, phone number or a username.

chpapa replied:
ok

chpapa commented on docs/specs/account-management-api.md:
```diff
@@ -0,0 +1,861 @@
+
+### Remove an email
+
+`DELETE /api/v1/account/identification`
```
I'm not sure are there still old browsers in use that don't support custom HTTP methods (sorry I know I'm too old lol), if there are we might want to make it possible for people to send a HTTP POST request with `method=DELETE`. 

If it is a 2000s things, 當我無講過 XD

louischan-oursky replied:
This method is well supported. https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods/DELETE

chpapa replied:
ok I'm cave man.

chpapa reviewed (commented):
@louischan-oursky A very quick glance over it.

louischan-oursky commented on docs/specs/account-management-api.md:
````diff
@@ -0,0 +1,861 @@
+```json
+{
+  "result": {
+    "identification_methods": [
````
I am not sure if identification is a countable noun. Some dictionaries I found on Google say it could be countable. 

In authflow, it is used as

```
{
  "identification": "email"
}
```

Let me change it to 

```
{
  "identifications": [
    {
      "identification": "email"
    }
  ]
}
```

louischan-oursky commented on docs/specs/account-management-api.md:
```diff
@@ -0,0 +1,861 @@
+    "code": 400,
+    "info": {
+      "cause": {
+        "kind": "DuplicatedIdentity"
```
The error object has the following fields. It has been used very long ago, and it is not invented by this spec.

- `name`: It is actually the category of the error.
- `code`: The HTTP status code of the error.
- `reason`: This is the main field we distinguish between different errors
- `info`: Extra information for a particular `reason`

The problem here is that we overload the reason `InvariantViolated` to represent some important errors. Opened another issue for this https://github.com/authgear/authgear-server/issues/4422

chpapa replied:
@louischan-oursky For #4422 can we actually try to improve our error object? 

1. Besides this API and Authflow, where is it being used?
2. I think instead of:

```json
"error": {
    "name": "Invalid",
    "reason": "InvariantViolated",
    "message": "identity already exists",
    "code": 400,
    "info": {
      "cause": {
        "kind": "DuplicatedIdentity"
      }
    }
}
```

We shall change to:
```
"error": {
  "type": "TypeOfError"
  "code": "Short String for the Error"
  "message": "Human readable message"
  "info": { }
}
```

The main changes from current design are:
- Use `type` instead of `name` for category of error. Name is very confusing
- HTTP status code should go back to HTTP response header
- `code` and `message` are the actual error
- We shall study the APIs used this error scheme, and try to come up with a standardized object for `info`, otherwise it is not usable at all (or else we need to write a spec to indicate what error will have which `info`, but it is not as good as a standardized object.

Let me copy this above message to #4422 as well.

louischan-oursky replied:
Continue the discussion in https://github.com/authgear/authgear-server/issues/4422

louischan-oursky commented on docs/specs/account-management-api.md:
````diff
@@ -0,0 +1,861 @@
+Request
+
+```json
+{
````
The endpoints of this API by definition are idempotent. For example, all the DELETE endpoints delete thing if they belong to the end-user. For update endpoints, it is also by definition idempotent, calling twice means no change. For those add endpoints, we cannot add duplicate identity by definition. So adding idempotency-key has little benefit.

chpapa replied:
In that case, I think maybe we should add in the first part of this spec, to define clearly being idemoptent is part of the commitment of all APIs (to be added there in future)

louischan-oursky commented on docs/specs/account-management-api.md:
```diff
@@ -0,0 +1,861 @@
+      "channel": "sms",
+      "otp_form": "code",
+      "code_length": 6,
+      "can_resend_at": "2023-09-21T00:00:00+08:00",
```
The developer needs to implement the cooldown counter in their UI. They need to know when they can trigger a resend so that they can show a counter.

chpapa replied:
ok

louischan-oursky commented on docs/specs/account-management-api.md:
```diff
@@ -0,0 +1,861 @@
+      "code_length": 6,
+      "can_resend_at": "2023-09-21T00:00:00+08:00",
+      "can_check": false,
+      "failed_attempt_rate_limit_exceeded": false
```
Yes, it is an error if the developer ignore this field. They will see an error with `reason: RateLimited`

louischan-oursky commented:
I updated the spec https://github.com/louischan-oursky/authgear-server/blob/dev-1267-specify-account-management-api/docs/specs/account-management-api.md

1. Removed the specification on the APIs we are not going to implement in near future.
2. Changed the path prefix to /api/v1
3. Added pseudo code to illustrate how the developer is supposed to integrate with the API

chpapa commented:
> I updated the spec https://github.com/louischan-oursky/authgear-server/blob/dev-1267-specify-account-management-api/docs/specs/account-management-api.md
> 
>     1. Removed the specification on the APIs we are not going to implement in near future.
> 
>     2. Changed the path prefix to /api/v1
> 
>     3. Added pseudo code to illustrate how the developer is supposed to integrate with the API

Thanks @louischan-oursky 

Random idea: do you think we should add a default `state` parameter in `authorization_url`, and check it at `query`, unless user explicitly turns it off? (Would it be "safer by default"? -- assume if users turn it off they know what they're doing.... but we make it default on)

BTW, do you think starting from now we should add a session on API spec for the error response? We don't have to approve it, but developers should add it to spec as they code and realize there might be some errors need to be thrown? (Of coz at design stage if foresee some errors already we can add it first)

louischan-oursky commented:
> do you think we should add a default state parameter in authorization_url, and check it at query, unless user explicitly turns it off?

I left this to the developer because the checking should be done when the developer app receives the callback. If state is generated by Authgear, then it has the same effect as `token` in the response.

> do you think starting from now we should add a session on API spec for the error response?

Yes. Should be part of https://linear.app/authgear/issue/DEV-1501/improve-error-response-in-apis to update existing specs to document what errors can be returned.

chpapa commented:
> I left this to the developer because the checking should be done when the developer app receives the callback. If state is generated by Authgear, then it has the same effect as token in the response.

But isn't that we can also check in the `query` of finish adding oauth provider as well? 

> Yes. Should be part of https://linear.app/authgear/issue/DEV-1501/improve-error-response-in-apis to update existing specs to document what errors can be returned.

I meant instead of waiting for the whole revamp, we shall start documenting the new error response for each API calls.

I saw that we already have documented `DuplicatedIdentity` in this APIs, but maybe we can have a table (as a standard format) for all new features? (btw we probably don't want to document ALL errors -- e.g. all API calls could have response an `Internal Server Error` but we don't need to keep repeating it -- but only certain type of error (hopefully we have a `type` / `name` that meant for application error. What's the possible `name` values we could have for it now?)

louischan-oursky commented:
> But isn't that we can also check in the query of finish adding oauth provider as well?

Yes. But the correct usage of `state` is to check the `state` parameter where the authorization code is received. In this case, it is the developer's app who receives the callback `?code=CODE&state=STATE`. In case of an attack, the developer checks `state`, then he then realizes this callback does not originate from the app. He then ignores the callback, and optional logs such incident. In a real world use case, the developer must use `state` to resume the flow, as shown in my pseudo code, otherwise, the developer has no idea what `token` to be used to resume the flow.

As the primary user of `state` is the developer, I want to leave it to him. `state` is really a parameter for the developer to resume the flow. Asking the developer to generate it makes him aware he is responsible for checking it when he receives the callback.

chpapa commented:
@louischan-oursky ummm, just wanna see if there are ways we can make it by default "safer", as developers would likely to forgot these sort of things. 

But on the other hand, in your sample code, I think the sample code also didn't really use `resume.state` for anything other than just matching `state` token? Is it really a bad thing if we enforce an issue and checking 

On the other hand, could we try to create the error table format that we can reuse both for this spec and another upcoming spec?

(Mainly is to decide the `name` (`type`) of error possible now.)

louischan-oursky commented:
@tung2744 Let's merge this?
